### DISCUSSION TITLE
How to properly type overload function to have correct inferred type from pyright?

### DISCUSSION THREAD

**Discussion by Holt59** [microsoft/pyright#8155] [Q&A] [answered]
I have the following code

```python
from typing import Literal, overload

@overload
def f1(model: str, convert: Literal[False] = ...) -> int: ...
@overload
def f1(model: str, convert: Literal[True] = ...) -> str: ...
@overload
def f1(model: str, convert: bool = ...) -> int | str: ...

def f1(model: str, convert: bool = True) -> int | str:
    ...

reveal_type(f1(""))  # int -> ko
reveal_type(f1("", convert=True))  # str -> okt
reveal_type(f1("", convert=False))  # int -> ok
```

Basically, I have two overloads for `f1` which takes a `bool` and the return type is different depending on the `bool` value. If I specify the `convert` parameter, `pyright` deduces the right type, but when I do not specify it, I think `pyright` considers the first overload even-though the default value is `bool`.

I guess it would be very complicated to extract the default value from the implementation, is there a way to have the right type deduce? I can think of two ways, but I am not sure if one is better:

- Put the `overload` with `Literal[True]` first.
- Had a third overload without the `convert`.

Is there a preferred? Maybe there is a better way then those two?

**Comment by erictraut** [comment, marked as answer]
You've specified three overloads with default values for `convert`. Only two of them should have a default value specified. The overload with `Literal[False]` should not include a `= ...` because `Literal[False]` is not a valid default value based on your implementation.

```python
@overload
def f1(model: str, convert: Literal[False]) -> int: ...
```

**Reply by Holt59** [reply]
Thanks!
